### PR TITLE
Move FP8 TE export logic to mcore.export

### PR DESCRIPTION
**Pull request by Laplasjan107**
# What does this PR do ?

Support FP8 TE TRT-LLM export with mcore path

**Collection**: llm/nlp

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
from nemo.export.tensorrt_llm import TensorRTLLM

trt_llm_exporter = TensorRTLLM(model_dir="/opt/checkpoints/tmp_triton_model_repository/")
trt_llm_exporter.export(nemo_checkpoint_path="/path/to/quantized/model", model_type="llama", fp8_quantized=True, fp8_kvcache=True)

# Autodetects when quantisation flags are not set
output = trt_llm_exporter.forward(["What is the best city in the world?"], max_output_token=17, top_k=1, top_p=0.0, temperature=1.0)
print("output: ", output)
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/export/tensorrt_llm.py [hunk 1]
```diff
@@ -83,6 +83,17 @@ def wrapper(*args, **kwargs):
     use_pytriton = False
 
 
+def determine_quantization_settings(
```
Comment (janekl) [inline nemo/export/tensorrt_llm.py:86]: Could you add docstrings for all new / edited functions?

Comment (janekl) [reply nemo/export/tensorrt_llm.py:86]: (same remark for typing)

File: nemo/export/tensorrt_llm.py [hunk 2]
```diff
@@ -353,17 +366,38 @@ def export(
                     from megatron.core.export.trtllm.trtllm_helper import TRTLLMHelper
                     from tensorrt_llm.layers import MoeConfig
 
+                    use_embedding_sharing = model_configs.get("share_embeddings_and_output_weights", False)
+                    fp8_quantized, fp8_kvcache = determine_quantization_settings(
+                        model_configs, fp8_quantized, fp8_kvcache
```
Comment (janekl) [inline nemo/export/tensorrt_llm.py:371]: It's `model_configs` (plural) passed, but only 1x `nemo_model_config` consumed in `determine_quantization_settings` - could we unify?

File: nemo/export/tensorrt_llm.py [hunk 3]
```diff
@@ -524,7 +559,7 @@ def get_transformer_config(self, nemo_model_config):
             ffn_hidden_size=nemo_model_config.get('ffn_hidden_size'),
             layernorm_epsilon=nemo_model_config.get('layernorm_epsilon'),
             add_bias_linear=nemo_model_config.get('bias'),
-            num_moe_experts=nemo_model_config.get('num_moe_experts', None),
+            num_moe_experts=num_moe_experts if num_moe_experts > 0 else None,
             normalization=transformer_config_normalization,
             layernorm_zero_centered_gamma=layernorm_zero_centered_gamma,
```
Comment (janekl) [inline nemo/export/tensorrt_llm.py:564]: In https://github.com/NVIDIA/NeMo/pull/11614 there is also `gated_linear_unit=nemo_model_config.get('gated_linear_unit'),` added.

* Can we add it here? So that we merge only this MR to offload CI slightly.
* Can we write `nemo_model_config.get('gated_linear_unit', False)` to conform with the typing? There could be suprises maybe, see `None` vs `0` for `num_moe_experts`

File: nemo/export/tensorrt_llm.py [hunk 4]
```diff
@@ -353,17 +366,38 @@ def export(
                     from megatron.core.export.trtllm.trtllm_helper import TRTLLMHelper
                     from tensorrt_llm.layers import MoeConfig
 
+                    use_embedding_sharing = model_configs.get("share_embeddings_and_output_weights", False)
```
Comment (janekl) [inline nemo/export/tensorrt_llm.py:369]: This overrides parameter `use_embedding_sharing` passed to `TensorRTLLM.export(...)`. Should it be removed from signature?

Comment (janekl) [reply nemo/export/tensorrt_llm.py:369]: Also: how about keeping the original name `share_embeddings_and_output_weights`?

File: nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py [hunk 5]
```diff
@@ -70,69 +70,68 @@ def __init__(self, path: Union[Path, TarPath]) -> None:
...
+    for key, value in state_dict.items():
+        if 'extra_state' not in key:
+            continue
 
-def load_scaling_factors(state_dict: dict, basename: str, size: int) -> Optional[torch.Tensor]:
-    keynames = [f'{basename}/shard_{layer}_{size}' for layer in range(size)]
-    bytes_list = [state_dict[keyname][0] for keyname in keynames]
-    return load_scales_from_bytes(bytes_list)
+        decomposed_sharded_key = key.split('/')
+        if not len(decomposed_sharded_key):
```
Comment (janekl) [inline nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py:114]: Always `len(decomposed_sharded_key) >= 1` so this `if` is inactive?

File: nemo/export/tensorrt_llm.py (hunk 1 above)
Comment (janekl) [reply nemo/export/tensorrt_llm.py]: If possible let's please use docstring convention from https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/api.py:

So sth like:
```
One or two sentences what's going on.

Args:
    param1 (Type): It is what it is.
    param2 (Type): It is what it is.

Returns:
    Type: It is what it is.
```

File: nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py [hunk 6]
```diff
@@ -77,69 +79,68 @@ def __init__(self, path: Union[Path, TarPath]) -> None:
...
 
-def standarize_distributed_scaling_factors(state_dict: dict) -> dict:
-    while key := get_extra_state_key(state_dict):
-        basename, size = unpack_extra_state_key(key)
-        scaling_factors = load_scaling_factors(state_dict, basename, size)
-        if scaling_factors is not None:
-            state_dict[basename + '.scale_fwd'] = scaling_factors
-        state_dict = clear_loaded_extra_states(state_dict, basename)
+        mcore_key = key_base.replace("layers", f"layers.{shard_layer}")
+        if isinstance(value, list) and len(value) == 1:
```
Comment (janekl) [inline nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py:138]: What are possible types of `value` if not only `List` (of different lengths)?

Comment (Laplasjan107) [reply nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py:138]: The extra states `value` is:
* a `BytesIO` object, when loaded from `zarr`
* a one-element list with a `BytesIO` object, when loaded from `torch_dist`

I'll add a comment here

File: nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py (hunk 5 above)
Comment (Laplasjan107) [reply nemo/export/trt_llm/nemo_ckpt_loader/nemo_file.py]: True, I've deleted this `if`

File: nemo/export/tensorrt_llm.py (hunk 3 above)
Comment (Laplasjan107) [reply nemo/export/tensorrt_llm.py:539]: Added it in the [latest commit](https://github.com/NVIDIA/NeMo/pull/11409/commits/732e73568f430a17871eb14558bba149a30347e9)

**Review by janekl — DISMISSED** [review]

**Review by janekl — APPROVED** [review]
